gef-ui/server/controllers/unissued-facilities/validation.js

```diff
@@ -0,0 +1,293 @@
+      msg += 'day ';
+      dateFieldsInError.push('coverEndDate-day');
+    }
+    if (!coverEndDateMonth) {
+      msg += !coverEndDateDay ? ' and month ' : 'month ';
+      dateFieldsInError.push('coverEndDate-month');
+    }
+    if (!coverEndDateYear) {
+      msg += !coverEndDateDay || !coverEndDateMonth ? 'and year' : 'year';
+      dateFieldsInError.push('coverEndDate-year');
+    }
+
+    aboutFacilityErrors.push({
+      errRef: 'coverEndDate',
+      errMsg: msg,
+      subFieldErrorRefs: dateFieldsInError,
+    });
+  }
+
+  if (body.facilityName && body.facilityName.length > 30) {
+    aboutFacilityErrors.push({
+      errRef: 'facilityName',
+      errMsg: 'Facility name cannot be more than 30 characters in length',
+    });
+  }
+
+  if (/[^A-Za-z0-9 .,:;'-]/.test(body.facilityName)) {
+    aboutFacilityErrors.push({
+      errRef: 'facilityName',
+      errMsg: 'Facility name must only include letters a to z, full stops, commas, colons, semi-colons, hyphens, spaces and apostrophes',
+    });
+  }
+
+  if (issueDateIsFullyComplete) {
+    issueDate = set(new Date(), { year: issueDateYear, month: issueDateMonth - 1, date: issueDateDay });
+  }
+
+  if (coverStartDateIsFullyComplete) {
+    coverStartDate = set(new Date(), { year: coverStartDateYear, month: coverStartDateMonth - 1, date: coverStartDateDay });
+  }
+
+  if (coverEndDateIsFullyComplete) {
+    coverEndDate = set(new Date(), { year: coverEndDateYear, month: coverEndDateMonth - 1, date: coverEndDateDay });
+  }
+
+  if (coverStartDateIsFullyComplete && coverEndDateIsFullyComplete) {
+    if (coverEndDate < coverStartDate) {
+      aboutFacilityErrors.push({
+        errRef: 'coverEndDate',
+        errMsg: 'Cover end date cannot be before cover start date',
+      });
+    }
+  }
+
+  if (issueDateIsFullyComplete && coverEndDateIsFullyComplete) {
+    if (coverEndDate < issueDate) {
+      aboutFacilityErrors.push({
+        errRef: 'coverEndDate',
+        errMsg: 'Cover end date cannot be before the issue date',
+      });
+    }
+  }
+
+  if (issueDateIsFullyComplete && coverStartDateIsFullyComplete) {
+    if (coverStartDate < issueDate) {
+      aboutFacilityErrors.push({
+        errRef: 'coverStartDate',
+        errMsg: 'Cover start date cannot be before the issue date',
+      });
+    }
+  }
+
+  // Regex tests to see if value is a number only
+  const digitsRegex = /^[0-9]*$/;
+  if (body.monthsOfCover) {
+    if (!digitsRegex.test(body.monthsOfCover)) {
+      aboutFacilityErrors.push({
+        errRef: 'monthsOfCover',
+        errMsg: 'You can only enter numbers',
+      });
+    }
+    if (body.monthsOfCover > 999) {
+      aboutFacilityErrors.push({
+        errRef: 'monthsOfCover',
+        errMsg: 'You can only enter a maximum of 999 months cover',
+      });
+    }
+  }
+
+  if (aboutFacilityErrors.length > 0) {
+    return res.render('partials/unissued-change-about-facility.njk', {
+      errors: validationErrorHandler(aboutFacilityErrors),
```

Comment:
I think it'd be better to have this validation file purely as validation and not passing or consuming req/res. So it just returns the validation errors - then in the actual route itself, call this function and do the res.render from there.

@abhi-markan thoughts?

Reply:
todo in future